pkg/operator/ceph/object/health.go

```diff
@@ -0,0 +1,269 @@
+/*
+Copyright 2020 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package object
+
+import (
+	"crypto/md5"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/pkg/errors"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	"github.com/rook/rook/pkg/clusterd"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	s3UserHealthCheckName      = "rook-ceph-internal-s3-user-checker"
+	s3HealthCheckBucketName    = "rook-ceph-internal-s3-bucket-checker"
+	defaultStatusCheckInterval = 1 * time.Minute
+	s3HealthCheckObjectBody    = "Test Rook Object Data"
+	s3HealthCheckObjectBodyMD5 = "5f286306a2227a156ba770800e71b796"
+	s3HealthCheckObjectKey     = "rookObj1"
+	contentType                = "plain/text"
+)
+
+// rgwStatusChecker aggregates the mon/cluster info needed to check the health of the monitors
+type rgwStatusChecker struct {
+	context        *clusterd.Context
+	objContext     *Context
+	interval       time.Duration
+	serviceIP      string
+	port           string
+	client         client.Client
+	namespacedName types.NamespacedName
+}
+
+// newRgwStatusChecker creates a new HealthChecker object
+func newRgwStatusChecker(context *clusterd.Context, objContext *Context, serviceIP, port string, client client.Client, namespacedName types.NamespacedName) *rgwStatusChecker {
+	c := &rgwStatusChecker{
+		context:        context,
+		objContext:     objContext,
+		interval:       defaultStatusCheckInterval,
+		serviceIP:      serviceIP,
+		port:           port,
+		namespacedName: namespacedName,
+		client:         client,
+	}
+
+	// allow overriding the check interval with an env var on the operator
+	checkInterval := os.Getenv("ROOK_CEPH_OBJECT_STATUS_CHECK_INTERVAL")
+	if checkInterval != "" {
+		if duration, err := time.ParseDuration(checkInterval); err == nil {
+			logger.Infof("ceph rgw status check interval is %q", checkInterval)
+			c.interval = duration
+		}
+	}
+
+	return c
+}
+
+// checkCephStatus periodically checks the health of the cluster
+func (c *rgwStatusChecker) checkObjectStore(stopCh chan struct{}) {
+
+	for {
+		select {
+		case <-stopCh:
+			// purge bucket and s3 user
+			// Needed for external mode where in converged everything goes away with the CR deletion
+			c.cleanupHealthCheck()
```

Comment:
Do we even need to worry about cleaning this up? If we're using the same bucket for all bucket tests in the converged or external clusters, then the bucket can exist for the lifetime of the object store. Or at least only the converged object store would need to delete the bucket during uninstall.

Reply:
External mode, external mode 🎶 . The cleanup ensures we always start fresh, e,g: if the operator crashes in the middle of the check, this could leave inconsistency behind and fail the subsequent PUT for instance. I believe it's safer.

Reply:
Agreed, per discussion we will have an independent bucket per external cluster so each one will have its own user creds as well.